scrapy/resolver.py

```diff
@@ -13,23 +14,27 @@
 @implementer(IResolverSimple)
 class CachingThreadedResolver(ThreadedResolver):
     """
-    Default caching resolver. IPv4 only, supports setting a timeout value for DNS requests
+    Default caching resolver. IPv4 only, supports setting a timeout value for DNS requests.
     """
 
-    @classmethod
-    def install(cls, reactor, settings):
-        if settings.getbool('DNSCACHE_ENABLED'):
-            cache_size = settings.getint('DNSCACHE_SIZE')
-        else:
-            cache_size = 0
-        resolver = cls(reactor, cache_size, settings.getfloat('DNS_TIMEOUT'))
-        reactor.installResolver(resolver)
-
     def __init__(self, reactor, cache_size, timeout):
         super(CachingThreadedResolver, self).__init__(reactor)
         dnscache.limit = cache_size
         self.timeout = timeout
 
+    @classmethod
+    def from_crawler(cls, crawler, reactor):
+        if crawler.settings.getbool('DNSCACHE_ENABLED'):
+            cache_size = crawler.settings.getint('DNSCACHE_SIZE')
+        else:
+            cache_size = 0
+        return cls(reactor, cache_size, crawler.settings.getfloat('DNS_TIMEOUT'))
+
+    @classmethod
+    def install_on_reactor(cls, reactor, crawler):
+        resolver = create_instance(cls, None, crawler, reactor=reactor)
```

Comment:
💄 `resolver = create_instance(cls, crawler=crawler, reactor=reactor)`.

Reply:
I know it's awkward, but [`create_instance`](https://github.com/scrapy/scrapy/blob/1.8.0/scrapy/utils/misc.py#L124) takes `settings` as a positional argument, that's why I'm explicitly passing `None`.
I guess we could also pass `crawler.settings` :man_shrugging:

Reply:
ah, you're right! passing None is fine then. we can also change crate_instance to make it less awkward, but it shouldn't be a part of this PR.